[17.0][ADD] pos_sale_color_code_tintometric_system, pos_sale_color_code_tintometric_system_link_pos_sale, pos_sale_color_code_tintometric_system_link_stock

IriaAlonso opened the pull request:
Add field `color_code_tintometric` on pos.order and frontend TPV

- [x] Show button on the product screen
- [x] Open form/wizard/something to add the color code tintometric
- [x] Confirm the product --> save at he product selected
- [x] Save the content in the backend
- [x] Show the content on the order receive view

- Add description

HT00479

IriaAlonso commented:
@dalonsod

IriaAlonso committed:
[ADD] pos_sale_color_code_tintometric_system
Add "Lot/S.Tin" field on POS order screen, POS receipt,
order POS on backend and order invoice

IriaAlonso committed:
[ADD] pos_sale_color_code_tintometric_system_link_pos_sale
Link between ‘pos_sale_color_code_tintometric_system’ and ‘pos_sale’
to ensure that when a sales order is transferred to the POS,
the 'Lot/S.Tin' field from pos.order.line is received.

DantePereyra commented on pos_sale_color_code_tintometric_system/static/src/app/utils/input_popups/textarea_popup.xml:
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Could you remove this to avoid unnecessary warnings?

IriaAlonso replied:
done :+1:

DantePereyra reviewed (changes requested):
Tested, LGTM. Minor comment.

dalonsod commented on pos_sale_color_code_tintometric_system/report/account_move_report_templates.xml:
```diff
@@ -0,0 +1,24 @@
+<odoo>
+    <template id="report_invoice_document" inherit_id="account.report_invoice_document">
+        <xpath expr="//th[@name='th_description']" position="after">
+            <t t-set="display_color" t-value="any(line.sudo().color_code_tintometric_pos for line in o.invoice_line_ids)"/>
```
Better use a more specific name for this variable, as current name seems confusing IMO:
```suggestion
            <t t-set="display_lot_stin" t-value="any(line.sudo().color_code_tintometric_pos for line in o.invoice_line_ids)"/>
```

dalonsod commented on pos_sale_color_code_tintometric_system/report/account_move_report_templates.xml:
```diff
@@ -0,0 +1,24 @@
+            </t>
+        </xpath>
+        <xpath expr="//td[@name='td_quantity']" position="before">
+            <t t-set="display_color_ln" t-value="any(line.sudo().color_code_tintometric_pos for line in o.invoice_line_ids)"/>
```
This seems to be wrong: if we show (or hide) column header if at least one line has Lot S/Tin filled, row cells should be shown depending on this as well. With current code, if no lines have Lot S/Tin, `<th>` is not shown, but `<td>` is available (empty anyway), obtaining more columns here.

So I think this variable should be removed and use generated below instead.

dalonsod commented on pos_sale_color_code_tintometric_system/report/account_move_report_templates.xml:
```diff
@@ -0,0 +1,24 @@
+        <xpath expr="//td[@name='td_quantity']" position="before">
+            <t t-set="display_color_ln" t-value="any(line.sudo().color_code_tintometric_pos for line in o.invoice_line_ids)"/>
+
+            <t t-if="display_color_ln">
```
See previous comment:
```suggestion
            <t t-if="display_lot_stin">
```

dalonsod commented on pos_sale_color_code_tintometric_system/report/stock_picking_report_templates.xml:
```diff
@@ -0,0 +1,46 @@
+<odoo>
+    <template id="report_delivery_document" inherit_id="stock.report_delivery_document">
+        <xpath expr="//table[@name='stock_move_line_table']//th[@name='th_sml_qty_ordered']" position="before">
+            <t t-set="display_color" t-value="any(move.sudo().move_color_code_tintometric_pos for move in o.move_ids_without_package)"/>
```
Regarding to variable name, make the same considerations than above

dalonsod commented on pos_sale_color_code_tintometric_system/models/stock_move.py:
```diff
@@ -0,0 +1,30 @@
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    move_pos_order_line_id = fields.Many2many("pos.order.line")
```
Due to usual `Many2many` conventions, rename this:
```suggestion
    move_pos_order_line_ids = fields.Many2many("pos.order.line")
```

dalonsod commented on pos_sale_color_code_tintometric_system/models/stock_move.py:
```diff
@@ -0,0 +1,30 @@
+    _inherit = "stock.move"
+
+    move_pos_order_line_id = fields.Many2many("pos.order.line")
+    move_color_code_tintometric_pos = fields.Char(compute="_compute_move_color_code_tintometric_pos", string="Lot/S.Tin pos")
```
Why not defining this as a simply `related` field?

dalonsod reviewed (changes requested):
Only code review, see comments

IriaAlonso committed:
[ADD] pos_sale_color_code_tintometric_system_link_stock
New 'Lot/S.Tin' field on Move Analysis filled from sale order line or pos order line

IriaAlonso commented:
@dalonsod

dalonsod commented on pos_sale_color_code_tintometric_system/models/stock_picking.py:
```diff
@@ -0,0 +1,21 @@
+                    lambda l: l.product_id == move.product_id
+                )
+
+                move.move_pos_order_line_id = [(6, 0, pos_lines.ids)]
```
@IriaAlonso this line fires an error when payment is done within POS frontend, because `move_pos_order_line_id` does not exist, possibly due to a field rename. Could you fix it and check if everything becomes OK with this field?

cc @lmiguens-solvos

IriaAlonso replied:
done
